src/postamble.js

```diff
@@ -311,17 +311,58 @@ function run(args) {
 Module['run'] = run;
 
 function exit(status, implicit) {
-  if (implicit && Module['noExitRuntime']) {
 #if ASSERTIONS
-    Module.printErr('exit(' + status + ') implicitly called by end of main(), but noExitRuntime, so not exiting the runtime (you can use emscripten_force_exit, if you want to force a true shutdown)');
+#if NO_EXIT_RUNTIME == 1
+  // compiler settings do not allow exiting the runtime, so flushing
+  // the streams is not possible. but in ASSERTIONS mode we check
+  // if there was something to flush, and if so tell the user they
+  // should request that the runtime be exitable.
+  // how we flush the streams depends on whether we are in NO_FILESYSTEM
+  // mode (which has its own special function for this; otherwise, all
+  // the code is inside libc)
+#if NO_FILESYSTEM
+  var flush = {{{ '$flush_NO_FILESYSTEM' in addedLibraryItems ? 'flush_NO_FILESYSTEM' : 'null' }}};
+#else
+  var flush = {{{ '$FS' in addedLibraryItems ? 'FS.quit' : "Module['_fflush']" }}};
 #endif
+  if (flush) {
```

Comment:
I don't quite understand why we are doing this dance with module.print. If it's ok to just call the users's `fflush`, why can't we just call it and let it flush for real?

Reply:
We don't normally even include `fflush` anymore, we only include it in `ASSERTIONS` builds. So we don't want to show the proper output here, we want to see if where would be a problem, had we not been in an `ASSERTIONS` build. So we just check if flushing would have emitting something (and we do so very carefully, as shutting down the FS may throw an error if it hit an error previously or something like that).

I can add more of an explanation to the code?

Reply:
Reading that again it does seem less than obvious, good point. Added a comment.